### PR TITLE
Add tests for sycl::use_kernel_bundle

### PR DESCRIPTION
**Pull request by vasilytric**
These tests do these verifications:
- exception was thrown in case of attempt to invoke kernel not contained in a bundle
- exception was thrown in case of primary context not equal to the kernel bundle's context
- exception was thrown in case of secondary context not equal to the kernel bundle's context
- exception was thrown in case, when when attempt to invoke kernel which incompatible with current tested device
- kernel from bundle should be invoked

____________________________
This PR depend on these PRs:
- https://github.com/KhronosGroup/SYCL-CTS/pull/218
- https://github.com/KhronosGroup/SYCL-CTS/pull/220
- https://github.com/KhronosGroup/SYCL-CTS/pull/224

### CODE REVIEW

**Review by ProGTX — APPROVED** [review]

File: tests/kernel_bundle/use_kernel_bundle.h [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  };
+};
+
+/** @brief Constructing empty kernel bundle
+ *  @param ctx Context that will be used for kernel bundle
+ *  @param num_bundles Number kernel bundles that will be constructed
+ */
+inline sycl::kernel_bundle<sycl::bundle_state::executable> get_empty_bundle(
+    const sycl::context &ctx) {
+  std::vector<sycl::kernel_id> kernel_ids{};
```
Comment (keryell) [inline tests/kernel_bundle/use_kernel_bundle.h:71]:
```suggestion
  std::vector<sycl::kernel_id> kernel_ids;
```
is the case by default

File: tests/kernel_bundle/use_kernel_bundle_call_set_spec_const_before_use_kb_no_second_queue.cpp [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+void run_verification(
+    sycl_cts::util::logger &log, const sycl::context &ctx,
+    sycl::kernel_bundle<sycl::bundle_state::executable> &kernel_bundle) {
+  sycl::queue queue(ctx, ctx.get_devices()[0]);
+  bool ex_was_thrown = false;
+  try {
+    queue.submit([&](sycl::handler &cgh) {
+      cgh.set_specialization_constant<spec_const>(0);
+      cgh.use_kernel_bundle(kernel_bundle);
+      cgh.single_task<kernel_for_kernel_bundle>([=]() {});
```
Comment (keryell) [inline tests/kernel_bundle/use_kernel_bundle_call_set_spec_const_before_use_kb_no_second_queue.cpp:39]:
```suggestion
      cgh.single_task<kernel_for_kernel_bundle>([=] {});
```

File: tests/kernel_bundle/use_kernel_bundle_call_set_spec_const_before_use_kb_no_second_queue.cpp [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    });
+  } catch (const sycl::exception &e) {
+    if (e.code() != sycl::errc::invalid) {
+      FAIL(log, unexpected_exception_msg);
+      throw;
+    }
+    ex_was_thrown = true;
```
Comment (keryell) [inline tests/kernel_bundle/use_kernel_bundle_call_set_spec_const_before_use_kb_no_second_queue.cpp:46]: I guess the code could be simplified if we add a global `check_exception_with_invalid` to write
```c++
  check_exception_with_invalid([&] {
    queue.submit([&](sycl::handler &cgh) {
      cgh.set_specialization_constant<spec_const>(0);
      cgh.use_kernel_bundle(kernel_bundle);
      cgh.single_task<kernel_for_kernel_bundle>([=] {});
    });
  });
```
for the full test.

File: tests/kernel_bundle/use_kernel_bundle_invoke_kernel_without_bundle_with_second_queue.cpp [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        "this bundle");
+    return;
+  }
+
+  bool ex_was_thrown = false;
+  try {
+    first_queue.submit(
+        [&](sycl::handler &cgh) {
+          cgh.use_kernel_bundle(kernel_bundle);
+          cgh.single_task<kernel>([=]() {});
```
Comment (keryell) [inline tests/kernel_bundle/use_kernel_bundle_invoke_kernel_without_bundle_with_second_queue.cpp:49]:
```suggestion
          cgh.single_task<kernel>([=] {});
```
all these can be simplified

File: tests/kernel_bundle/use_kernel_bundle_second_queue_ctx_not_equal_to_kernel_bundle_ctx.cpp [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    set_test_info(out, TOSTRING(TEST_NAME), TEST_FILE);
+  }
+
+  /** execute the test
+   */
+  void run(util::logger &log) override {
+    auto platforms = sycl::platform::get_platforms();
+    const std::vector<sycl::device> devices{sycl::device::get_devices()};
+    if (devices.size() <= 1) {
+      log.note("Test skipped due to one device was found");
```
Comment (keryell) [inline tests/kernel_bundle/use_kernel_bundle_second_queue_ctx_not_equal_to_kernel_bundle_ctx.cpp:42]: There was a comment on this on another PR

File: tests/kernel_bundle/use_kernel_bundle_verify_that_kernel_was_invoked.cpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  This test obtain kernel bundle in bundle_state::input, then for this kernel
+//  calls set_spec_const with some different value, then call sycl::build to
+//  have kernel bundle in bundle_state::executable state and calls
+//  sycl::handler::use_kernel_bundle for this bundle, then get spec const value.
+//
+//  The test verifies that the pec const value is equal to expected velue.
```
Comment (keryell) [inline tests/kernel_bundle/use_kernel_bundle_verify_that_kernel_was_invoked.cpp:10]:
```suggestion
//  The test verifies that the specialization constant value is equal to expected value.
```

**Review by keryell — COMMENTED** [review]
Good!
But there are some easy factorizing for the exception testing which could shorten some tests.

File: tests/kernel_bundle/use_kernel_bundle_verify_that_kernel_was_invoked.cpp (hunk 6 above)
Comment (vasilytric) [reply tests/kernel_bundle/use_kernel_bundle_verify_that_kernel_was_invoked.cpp]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/225/commits/e157b8bd72174f5b18b4e387b9dfb4e353573bd7#diff-d950e7de8902d28e945f2971a8e619986ed79fcf9076a4205047beac0da4175aL10-L11)

File: tests/kernel_bundle/use_kernel_bundle_second_queue_ctx_not_equal_to_kernel_bundle_ctx.cpp (hunk 5 above)
Comment (vasilytric) [reply tests/kernel_bundle/use_kernel_bundle_second_queue_ctx_not_equal_to_kernel_bundle_ctx.cpp]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/225/commits/e157b8bd72174f5b18b4e387b9dfb4e353573bd7#diff-d4520edbb0ca7bc678e694703cae85b14192ee2dd0d8ce877de37be0d7668537L42-L43)

File: tests/kernel_bundle/use_kernel_bundle_invoke_kernel_without_bundle_with_second_queue.cpp (hunk 4 above)
Comment (vasilytric) [reply tests/kernel_bundle/use_kernel_bundle_invoke_kernel_without_bundle_with_second_queue.cpp:49]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/225/commits/e157b8bd72174f5b18b4e387b9dfb4e353573bd7#diff-2e891a7394b6c8dfaf6562b100ea5ecc221d6aec61281549a988fe39eda86283L56-L57)

File: tests/kernel_bundle/use_kernel_bundle_call_set_spec_const_before_use_kb_no_second_queue.cpp (hunk 3 above)
Comment (vasilytric) [reply tests/kernel_bundle/use_kernel_bundle_call_set_spec_const_before_use_kb_no_second_queue.cpp:46]: @keryell It's a good idea, but in some cases we need call `set_specialization_constant` before calling `use_kernel_bundle`, in some cases we need do it after calling `use_kernel_bundle`.
In some cases we should use the secondary `queue`.

For all these cases we should have a few different `check_exception_with_invalid`. But in current solution we have `run_verification` function that  contains all things that will be done in this tests, but implementing a `check_exception_with_invalid` can add one encapsulation level that will hide some test actions. Therefore, I suggest do not change current test structure.

File: tests/kernel_bundle/use_kernel_bundle_call_set_spec_const_before_use_kb_no_second_queue.cpp (hunk 2 above)
Comment (vasilytric) [reply tests/kernel_bundle/use_kernel_bundle_call_set_spec_const_before_use_kb_no_second_queue.cpp]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/225/commits/e157b8bd72174f5b18b4e387b9dfb4e353573bd7#diff-d25bca94ae123b3ee37637a07382213babede7e443b70583f54eedd749ceec2dL39-L40) and in other cases

File: tests/kernel_bundle/use_kernel_bundle.h (hunk 1 above)
Comment (vasilytric) [reply tests/kernel_bundle/use_kernel_bundle.h]: I've applied this comment [here](https://github.com/KhronosGroup/SYCL-CTS/pull/225/commits/e157b8bd72174f5b18b4e387b9dfb4e353573bd7#diff-64ba6ecf294608e00a3e4b708436967084a1c9631cf800f444d742eea97485afL71-L72)

File: tests/kernel_bundle/use_kernel_bundle_call_set_spec_const_before_use_kb_no_second_queue.cpp (hunk 3 above)
Comment (keryell) [reply tests/kernel_bundle/use_kernel_bundle_call_set_spec_const_before_use_kb_no_second_queue.cpp:46]: OK. You know better your code than me. :-)

**Review by keryell — APPROVED** [review]
Thanks!

**Comment by vasilytric** [comment]
I am going to merge `SYCL-2020` branch into this branch after PR#224 will be merged because of the fact that this PR is depend on PR#224

**Comment by bader** [comment]
Doesn't compile with dpcpp due to:
```
In file included from ../tests/kernel_bundle/use_kernel_bundle_call_set_spec_const_after_use_kb_no_second_queue.cpp:13:
../tests/kernel_bundle/use_kernel_bundle.h:61:5: error: use of undeclared identifier 'm_print_bundle_state'
     m_print_bundle_state = false;
     ^
```

**Comment by vasilytric** [comment]
> Doesn't compile with dpcpp due to:

@bader It's a strange error that is not reproduced locally, but adding `this->` resolved the error. All checks have passed now.

**Pull request merged**